dbfit-java/derby/src/main/java/dbfit/environment/DerbyEnvironment.java

```diff
@@ -82,9 +84,92 @@ public Pattern getParameterPattern() {
         }
     }
 
+    private boolean isProcedureOrFunction(String schemaName, String objectName,
+                                                             boolean checkForFunction)
+                                                                   throws SQLException {
+        DatabaseMetaData dbmd = getConnection().getMetaData();
+        try (ResultSet rs = checkForFunction ? dbmd.getFunctions(null, schemaName, objectName) :
+                                               dbmd.getProcedures(null, schemaName, objectName)) {
+            return rs.next();
+        }
+    }
+
+    private Direction getParamDirection(int columnType, boolean isFunction) {
+        if (isFunction) {
+            switch (columnType) {
+                case DatabaseMetaData.functionColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.functionReturn:
+                    return Direction.RETURN_VALUE;
+            }
+        } else {
+            switch (columnType) {
+                case DatabaseMetaData.procedureColumnIn:
+                    return Direction.INPUT;
+                case DatabaseMetaData.procedureColumnInOut:
+                    return Direction.INPUT_OUTPUT;
+                case DatabaseMetaData.procedureColumnOut:
+                    return Direction.OUTPUT;
+            }
+        }
+        throw new UnsupportedOperationException("Type " + columnType
+                      + " is not a supported type for "
+                      + (isFunction ? "function" : "procedure") + " parameters");
+    }
+
+    private Map<String, DbParameterAccessor> getParamMap(String schemaName,
+                                                         String procedureName,
+                                                         boolean isFunction)
+                                                         throws SQLException {
+        DatabaseMetaData dbmd = getConnection().getMetaData();
+        int position = 0;
+        Direction paramDirection;
+        String dataType, paramName;
+        Map<String, DbParameterAccessor> allParams = new HashMap<>();
+        try (ResultSet rs = isFunction ? dbmd.getFunctionColumns(null, schemaName, procedureName, "%") :
```

Comment:
Seems a diverging thought at a bit lower level of detail. But it's just a point which made me wonder, not necessarily a good idea. Another stretch (possibly one going too far) around that could be swapping the if/conditional with polymorphic hierarchy.

In general my wondering was mostly about [SLAP, SRP, DRY principles](http://nithinbekal.com/notes/design-principles/). Which have their pros but cons too. So it's up to experimentation and to your judgement.